### ISSUE TITLE
Performance: Updated benchmarks and CPU Platform

### ISSUE THREAD

**Issue by zonca** [paesanilab/mbpol_openmm_plugin#119]
## Updated benchmarks

I noticed some inconsistencies in the benchmarks, so I reran all of them more accurately. I believe by mistake I was running AMOEBA with OpenMM compiled in Debug mode, so it was 10x slower than expected in some cases. So the performance numbers for AMOEBA changed, mbpol remained the same.

Here are all the new benchmarks (click on "Revisions" to compare with older):

https://gist.github.com/zonca/e41cd831afcc28f99638f7acef8f7d15

Now in the Reference platform, mbpol is 20x slower than AMOEBA, this might be due to the fact that we derived Electrostatics from a older version of the AMOEBA plugin. The new plugin has a newer version of the dipole converging algorithm (named DIIS) that might be the cause of the performance difference (needs more investigation).

## CPU Platform: multi-threaded induced dipoles

I had implemented multi-threading in the dipole loop, but yesterday I quoted wrong timings in the chat. It actually didn't provide any performance improvements. This was due to the fact that there is a loop through pairs of particles that I parallelized with threads to compute the induced dipole field in parallel. However this parallel computation needs to be reduced back to a single array at each iteration of the dipole convergence loop. Unfortunately there is not enough speed gain in the parallel section to offset the reduction (even if I also implemented the reduction in parallel, i.e. I created `numThreads` arrays of the field vector, compute them in parallel, once completed, made another parallel loop over particles to sum all the `numThreads` vectors).

## CPU Platform: multi-threaded electrostatics computation

So I reverted back the induced dipole iterations to a single-threaded code and implemented multi-threading in the following section where the actual electrostatics interaction is computed. In this section there is much more computing per thread and there is a significant speedup, this provides about a 4-6x speedup using 12 threads in this section alone.
This reflects in a 2-3x speedup in the complete computation, see the benchmarks: https://gist.github.com/zonca/e41cd831afcc28f99638f7acef8f7d15
I do not think that without a major rewrite we can gain much more on the CPU platform for Electrostatics. But I still need to implement multi-threading in PME.

## Future work

There are 3 directions we could follow to improve performance further:

* Implement multi-threading in the PME version of Electrostatics [~1/2 weeks]
* Investigate more the performance difference between AMOEBA and MBPol and if useful implement the new dipole convergence algorithm [1 week for investigation, ~2/3 weeks for implementation]
* Implement multi-threading in 2 and 3 body [~3/5 weeks for implementation]

Do you have any questions? What should we do next? (if easier we can talk next wed)

**Comment by zonca** [comment]
completed the implementation of PME electrostatics, performance improvement is minimal because most of the time is spent in the dipole convergence loop instead of the electrostatics loop for PME. https://gist.github.com/zonca/e41cd831afcc28f99638f7acef8f7d15

Next I'll work on the second step outlined above

**Comment by gmedders** [comment]
Hey @zonca, it's been a long time! Apparently my notifications were going to my defunct ucsd email. I just changed my notifications, so sorry if I missed any tags.

Just in case you're not already aware, I just wanted to point out that in the DLPOLY implementation, we typically used the time-reversible always stable predictor corrector method ([ASPC](http://onlinelibrary.wiley.com/doi/10.1002/jcc.10385/abstract)) method for the induced dipoles during actual simulations, rather than iteration. In this method, while you still have to iteratively converge the induced dipoles during the first simulation step, the dipoles for all subsequent steps are obtained by:

 - Keeping a history of the induced dipoles for the past N steps (typically, N=4)
 - Extrapolating the induced dipole of the current step from the past N steps
 - "Correcting" the extrapolation (equivalent to single step through the iteration loop, rather than looping until convergence)

Thus (for all but the first simulation step), this should naively make calculating the induced dipole roughly as expensive as calculating the permanent electrostatics (and will make the performance improvement from your PME implementation much more apparent!). And, in that sense, measuring (serial/parallel) efficiency based on a single energy calculation with iteratively-solved dipoles is somewhat less meaningful, since this doesn't mimic the typical use case during MD.

I'm sure @chemphys or someone can point you to where this is implemented in DL_POLY if it's useful (sorry, Marc, for volunteering you, haha). Or if no one can find it, I can also help out ;)

Hope all is well

**Comment by zonca** [comment]
thanks @gmedders !

Currently I am using the iterative loop, which is now named `convergeInduceDipolesBySOR` in the AMOEBA plugin: https://github.com/pandegroup/openmm/blob/7.1_branch/plugins/amoeba/platforms/reference/src/SimTKReference/AmoebaReferenceMultipoleForce.cpp#L933

This algorithm is not used anymore in the AMOEBA plugin, instead, depending on polarization type, they use:

* if polarization type is Mutual (the default value), they use DIIS: https://github.com/pandegroup/openmm/blob/7.1_branch/plugins/amoeba/platforms/reference/src/SimTKReference/AmoebaReferenceMultipoleForce.cpp#L1014
* if polarization type is Extrapolated, they use converge by Extrapolation: https://github.com/pandegroup/openmm/blob/7.1_branch/plugins/amoeba/platforms/reference/src/SimTKReference/AmoebaReferenceMultipoleForce.cpp#L966

@agoetz @fpaesani would it be useful to implement one of those algorithms in the plugin? or ASPC as suggested by Greg?

**Comment by fpaesani** [comment]
Yes! Anything that can speed up the code and scale it to large number of processors is key. I would suggest to start with DIIS.

**Comment by agoetz** [comment]
I second Francesco's opinion. I would go with DIIS since replicating code within OpenMM is probably less work.
Is it also available for the GPU platform?
DIIS requires matrix multiplications. Do you know the rank and how it scales with particle numbers? Matrix multiplications are efficient on GPUs but it might make the algorithm expensive for large systems.

**Comment by zonca** [comment]
@agoetz yes, it is available for GPU as well https://github.com/pandegroup/openmm/blob/master/plugins/amoeba/platforms/cuda/src/kernels/multipoleInducedField.cu#L605

I don't know any details about scaling.

**Comment by zonca** [comment]
let's pause further development on this until we have taken a decision if using Marc's code is better
